#Q1. Pearson correlation coefficient is a measure of the linear relationship between two variables. Suppose
you have collected data on the amount of time students spend studying for an exam and their final exam
scores. Calculate the Pearson correlation coefficient between these two variables and interpret the result.

In [63]:
## example dataset: using scores dataset from kaggle
import pandas as pd
df = pd.read_csv("services.csv")
df.head()

id  location_id  program_id accepted_payments alternate_name  \
0   1            1         NaN               NaN            NaN   
1   2            2         NaN               NaN            NaN   
2   3            3         NaN               NaN            NaN   
3   4            4         NaN               NaN            NaN   
4   5            5         NaN               NaN            NaN   

                           application_process  \
0                   Walk in or apply by phone.   
1           Apply by phone for an appointment.   
2  Phone for information (403-4300 Ext. 4322).   
3                              Apply by phone.   
4                       Phone for information.   

                                            audience  \
0  Older adults age 55 or over, ethnic minorities...   
1       Residents of San Mateo County age 55 or over   
2  Older adults age 55 or over who can benefit fr...   
3  Parents, children, families with problems of c...   
4  Low-income working families with children tran...   

                                         description  \
0  A walk-in center for older adults that provide...   
1  Provides training and job placement to eligibl...   
2  Offers supportive counseling services to San M...   
3  Provides supervised visitation services and a ...   
4  Provides fixed 8% short term loans to eligible...   

                                         eligibility email  ...  \
0  Age 55 or over for most programs, age 60 or ov...   NaN  ...   
1  Age 55 or over, county resident and willing an...   NaN  ...   
2        Resident of San Mateo County age 55 or over   NaN  ...   
3                                               None   NaN  ...   
4  Eligibility: Low-income family with legal cust...   NaN  ...   

  interpretation_services                                           keywords  \
0                     NaN  ADULT PROTECTION AND CARE SERVICES, Meal Sites...   
1                     NaN  EMPLOYMENT/TRAINING SERVICES, Job Development,...   
2                     NaN  Geriatric Counseling, Older Adults, Gay, Lesbi...   
3                     NaN  INDIVIDUAL AND FAMILY DEVELOPMENT SERVICES, Gr...   
4                     NaN     COMMUNITY SERVICES, Speakers, Automobile Loans   

  languages                               name required_documents  \
0       NaN    Fair Oaks Adult Activity Center                NaN   
1       NaN   Second Career Employment Program                NaN   
2       NaN             Senior Peer Counseling                NaN   
3       NaN           Family Visitation Center                NaN   
4       NaN  Economic Self-Sufficiency Program                NaN   

      service_areas  status wait_time website taxonomy_ids  
0             Colma  active  No wait.     NaN          NaN  
1  San Mateo County  active   Varies.     NaN          NaN  
2  San Mateo County  active   Varies.     NaN          NaN  
3  San Mateo County  active  No wait.     NaN          NaN  
4  San Mateo County  active       NaN     NaN          NaN  

[5 rows x 22 columns]

In [64]:
# Calculating the pearson correlation
df.corr(method="pearson")

/tmp/ipykernel_3785/4007322474.py:2: FutureWarning: The default value of numeric_only in DataFrame.corr is deprecated. In a future version, it will default to False. Select only valid columns or specify the value of numeric_only to silence this warning.
  df.corr(method="pearson")


id  location_id  program_id
id           1.000000     0.999578         NaN
location_id  0.999578     1.000000         NaN
program_id        NaN          NaN         NaN

The pearson correlation coeffecient between Scores and Study hours is 0.97 which indicates a strong positive correlation between the two varaibles. An increase in study hours will lead to an increase in scores

#Q2. Spearman's rank correlation is a measure of the monotonic relationship between two variables.
Suppose you have collected data on the amount of sleep individuals get each night and their overall job
satisfaction level on a scale of 1 to 10. Calculate the Spearman's rank correlation between these two
variables and interpret the result.

In [65]:
## example dataset
import pandas as pd
df = pd.DataFrame({
    "sleep" : [8,7,6,8,9,3,4,3],
    "satisfaction" : [9,8,8,6,8,3,4,3]
})
df.head()

sleep  satisfaction
0      8             9
1      7             8
2      6             8
3      8             6
4      9             8

In [66]:
## Spearman correlation
df.corr(method="spearman")

sleep  satisfaction
sleep         1.000000      0.782745
satisfaction  0.782745      1.000000

The spearman rank correlation coeffecient between sleep hours and job satisfaction is 0.783 which indicates a strong positive correlation between the two varaibles. An increase in sleep hours will lead to an increase in job satisfaction

#Q3. Suppose you are conducting a study to examine the relationship between the number of hours of
exercise per week and body mass index (BMI) in a sample of adults. You collected data on both variables
for 50 participants. Calculate the Pearson correlation coefficient and the Spearman's rank correlation
between these two variables and compare the results.

In [67]:
# example dataset
import pandas as pd
import numpy as np
df = pd.DataFrame({
    "exercise" : np.random.randint(1,10,50),
    "bmi" : np.random.randint(15,30,50)
})
df.head()

exercise  bmi
0         8   21
1         2   27
2         1   22
3         3   20
4         1   20

In [68]:
#pearson correlation
df.corr()

exercise      bmi
exercise   1.00000 -0.00787
bmi       -0.00787  1.00000

In [69]:
#spearman rank correlation
df.corr(method="spearman")

exercise       bmi
exercise  1.000000 -0.003723
bmi      -0.003723  1.000000

The pearson correlation coeffecient for exercise and bmi = 0.082
The spearman rank correlation coeffecient for exercise and bmi = 0.076
In this above example the pearson correlation coeffecient showed marginally higher values compared to spearman rank correlation coeffecient and since both coeffecients are positive and very close to zero we can infer that from the dataset it appears that exercise and bmi have a very weak positive relation if any at best.

#Q4. A researcher is interested in examining the relationship between the number of hours individuals
spend watching television per day and their level of physical activity. The researcher collected data on
both variables from a sample of 50 participants. Calculate the Pearson correlation coefficient between
these two variables.

In [70]:
import pandas as pd
import numpy as np
df = pd.DataFrame({
    "tv" : np.random.randint(1,12,50),
    "activity": np.random.randint(1,4,50)
})

In [71]:
df.corr()

tv  activity
tv        1.000000 -0.078905
activity -0.078905  1.000000

In [72]:
df.corr(method="spearman")

tv  activity
tv        1.000000 -0.106094
activity -0.106094  1.000000

The pearson correlation coeffecient for v watching per day and activity = -0.079 The spearman rank correlation coeffecient for tv watching per day and activity = -0.129281 In this above example the pearson correlation coeffecient showed a higher value compared to spearman rank correlation coeffecient and since both coeffecients are negative and closer to zero we can infer that from the dataset it appears that watching tv is weakly correlates negatively with activity and as number of hours of tv watching increase, the physical activity hours are decreasing. Spearman coeffecient infers a higher negative correlation when compared to pearson coeffecient

#Q5. A survey was conducted to examine the relationship between age and preference for a particular
brand of soft drink. The survey results are shown below:Age(Years)
25 Coke
42 Pepsi
37 Mountain dew
19 Coke
31 Pepsi
28 Coke






In [73]:
#create df
import pandas as pd
df = pd.DataFrame({
    "Age(Years)" : [25,42,37,19,31,28],
    "Soft drink Preference" : ["Coke","Pepsi","Mountain dew","Coke","Pepsi","Coke"]
})

In [74]:
#using target guided ordinal encoding
mean_age = df.groupby("Soft drink Preference")["Age(Years)"].mean().to_dict()

In [75]:
df["soft_drink_encoded"]  = df["Soft drink Preference"].map(mean_age)

In [76]:
df

Age(Years) Soft drink Preference  soft_drink_encoded
0          25                  Coke                24.0
1          42                 Pepsi                36.5
2          37          Mountain dew                37.0
3          19                  Coke                24.0
4          31                 Pepsi                36.5
5          28                  Coke                24.0

In [77]:
df[["Age(Years)","soft_drink_encoded"]].corr()

Age(Years)  soft_drink_encoded
Age(Years)             1.00000             0.83753
soft_drink_encoded     0.83753             1.00000

In [78]:
df[["Age(Years)","soft_drink_encoded"]].corr(method="spearman")

Age(Years)  soft_drink_encoded
Age(Years)            1.000000            0.833238
soft_drink_encoded    0.833238            1.000000

mean age of soft drink preference and age have a pearson correlation coeffecient of 0.837 and a spearman correlation of 0.833 both of which indicate a strong positive correlation wherein increase in age will also result in an change(increase) in the mean age of soft drink preference (from coke to mountain dew)

#Q6. A company is interested in examining the relationship between the number of sales calls made per day
and the number of sales made per week. The company collected data on both variables from a sample of
30 sales representatives. Calculate the Pearson correlation coefficient between these two variables.

In [79]:
#random df
import pandas as pd
import numpy as np
df = pd.DataFrame({
    "sales_calls_per_day" : np.random.randint(1,51,30),
    "sales_per_week": np.random.randint(11,61,30)
})

In [80]:
df.corr() #pearson correlation

sales_calls_per_day  sales_per_week
sales_calls_per_day             1.000000        0.125293
sales_per_week                  0.125293        1.000000

The pearson correlation for sales calls per day and number of sales per week is 0.340796 which indicates a weak positive correlation